guillotina_cms/api/content.py

```diff
@@ -17,3 +27,140 @@
+
+    # verify current order matches
+    txn = get_transaction()
+    if not IPostgresStorage.providedBy(txn.storage) or ICockroachStorage.providedBy(txn.storage):
+        raise HTTPBadRequest(content={
+            'message': 'Content ordering not supported'
+        })
+
+    conn = await txn.get_connection()
+    results = await conn.fetch('''
+select id, (json->>'position_in_parent')::int as pos from {}
+WHERE parent_id = $1 AND of IS NULL
+ORDER BY (json->>'position_in_parent')::int ASC
+limit {}'''.format(
+        txn.storage._objects_table_name, MAX_FOLDER_SORT_SIZE), context._p_oid)
+    if len(results) >= MAX_FOLDER_SORT_SIZE:
+        raise HTTPPreconditionFailed(content={
+            'message': 'Content ordering not supported on folders larger than {}'.format(
+                MAX_FOLDER_SORT_SIZE
+            )
+        })
+
+    results.sort(key=lambda item: item['pos'] or 0)
+    order = []
+    mapped = {}
+    for item in results:
+        order.append(item['id'])
+        mapped[item['id']] = item['pos']
+
+    if len(subset_ids) > len(order):
+        raise HTTPPreconditionFailed(content={
+            'message': 'Invalid subset. More values than current ordering'
+        })
+    if len(subset_ids) == len(order):
+        if subset_ids != order:
+            raise HTTPPreconditionFailed(content={
+                'message': 'Invalid subset',
+                'current': order
+            })
+    else:
+        # verify subset
+        # find current ordered subset
+        try:
+            start = order.index(subset_ids[0])
+            end = order.index(subset_ids[-1]) + 1
+        except ValueError:
+            raise HTTPPreconditionFailed(content={
+                'message': 'Invalid subset. Could not calculate subset match',
+
+            })
+        order_subset = order[start:end]
+        if subset_ids != order_subset:
+            raise HTTPPreconditionFailed(content={
+                'message': 'Invalid subset',
+                'current': order_subset
+            })
+
+    if ((order.index(data['obj_id']) + delta + 1) > len(order) or (
+            order.index(data['obj_id']) + delta) < 0):
+        raise HTTPPreconditionFailed(content={
+            'message': 'Can not move. Invalid move target.'
+        })
+    # now swap position for item
+    moved_item_index = order.index(data['obj_id'])
+    moved = {}
+    # over range of delta and shift position of the rest the opposite direction
+    # for example:
+    #  - move idx 0, delta 3
+    #    - idx 1, 2, 3 are moved to 0, 1, 2
+    #  - move idx 4, delta -2
+    #    - idx 2, 3 are moved to 3, 4
+    if delta < 0:
+        group = [i for i in reversed(
+            order[moved_item_index + delta:moved_item_index + 1])]
+    else:
+        group = order[moved_item_index:moved_item_index + delta + 1]
+
+    for item_id in group[1:]:
+        moved.update(
+            await _swap(context, data['obj_id'], item_id, order, mapped))
+
+    return moved
```

Comment:
Maybe we could split on different functions, as its really a long one

Reply:
How about making the endpoint of class?